File: Mika/trunk/core-vm/java/java/io/ObjectStreamField.java
/**************************************************************************
* Copyright  (c) 2001 by Acunia N.V. All rights reserved.                 *
*                                                                         *
* This software is copyrighted by and is the sole property of Acunia N.V. *
* and its licensors, if any. All rights, title, ownership, or other       *
* interests in the software remain the property of Acunia N.V. and its    *
* licensors, if any.                                                      *
*                                                                         *
* This software may only be used in accordance with the corresponding     *
* license agreement. Any unauthorized use, duplication, transmission,     *
*  distribution or disclosure of this software is expressly forbidden.    *
*                                                                         *
* This Copyright notice may not be removed or modified without prior      *
* written consent of Acunia N.V.                                          *
*                                                                         *
* Acunia N.V. reserves the right to modify this software without notice.  *
*                                                                         *
*   Acunia N.V.                                                           *
*   Vanden Tymplestraat 35      info@acunia.com                           *
*   3000 Leuven                 http://www.acunia.com                     *
*   Belgium - EUROPE                                                      *
**************************************************************************/

/*
** $Id: ObjectStreamField.java,v 1.2 2006/03/29 09:27:14 cvs Exp $
*/

package java.io;

import java.lang.reflect.Field;

public class ObjectStreamField implements Comparable {

  Class type;
  char code;
  String name;
  String typeString;
  boolean isPrimitive;

  /** will only be set by the ObjectStreamClass when needed ! */
  Field field;
  private boolean unshared;

  private native void create(String name, Class type);

  public ObjectStreamField (String name, Class type) {
    create(name, type);
  }

  public ObjectStreamField (String name, Class type, boolean unshared) {
    this.unshared = unshared;
    create(name, type);
  }

  ObjectStreamField(String name, String desc, char code){
    this.name = name;
    typeString = desc;
    this.code = code;
  }

  public boolean isUnshared() {
    return unshared; 
  }
  
  public String getName() {
    return this.name;
  }
  
  public Class getType() {
    return this.type;
  }

  public char getTypeCode() {
    return this.code;
  }

  public String getTypeString(){
    return typeString;
  }

  public boolean isPrimitive() {
    return isPrimitive;
  }

  public String toString() {
    return "ObjectStreamField: name = '"+name+"' of "+type;
  }

  public int compareTo (Object o) {
    ObjectStreamField target = (ObjectStreamField)o;
    boolean p = isPrimitive;
    int answer;
    if (p != target.isPrimitive) {
      answer =  (p ? -1 : 1 );
    }
    else {
     answer = name.compareTo(target.name);
    }
    return answer;
  }

  //TODO ...

  public int getOffset(){
    System.out.println("method java.io.ObjectStreamField.getOffset() is not implemented");
    return -1;
  }

  protected void setOffset(int off){
    System.out.println("method java.io.ObjectStreamField.setOffset(int) is not implemented");
  }


}

File: Mika/tags/RELEASE-1.4.1/core-vm/src/native/java/io/ObjectStreamField.c
/**************************************************************************
* Copyright (c) 2001, 2002, 2003 by Acunia N.V. All rights reserved.      *
*                                                                         *
* This software is copyrighted by and is the sole property of Acunia N.V. *
* and its licensors, if any. All rights, title, ownership, or other       *
* interests in the software remain the property of Acunia N.V. and its    *
* licensors, if any.                                                      *
*                                                                         *
* This software may only be used in accordance with the corresponding     *
* license agreement. Any unauthorized use, duplication, transmission,     *
*  distribution or disclosure of this software is expressly forbidden.    *
*                                                                         *
* This Copyright notice may not be removed or modified without prior      *
* written consent of Acunia N.V.                                          *
*                                                                         *
* Acunia N.V. reserves the right to modify this software without notice.  *
*                                                                         *
*   Acunia N.V.                                                           *
*   Philips Site 5, box 3       info@acunia.com                           *
*   3001 Leuven                 http://www.acunia.com                     *
*   Belgium - EUROPE                                                      *
*                                                                         *
* Modifications copyright (c) 2004, 2005, 2006 by Chris Gray,             *
* /k/ Embedded Java Solutions. All rights reserved.                       *
*                                                                         *
**************************************************************************/

/*
** $Id: ObjectStreamField.c,v 1.4 2006/05/23 14:54:14 cvs Exp $
*/

#include <string.h>

#include "core-classes.h"
#include "clazz.h"
#include "exception.h"
#include "fields.h"
#include "heap.h"
#include "loading.h"
#include "wstrings.h"
#include "threads.h"
#include "ts-mem.h"
#include "descriptor.h"

void ObjectStreamField_create(JNIEnv *env, w_instance this, w_instance String, w_instance Type) {

  w_thread thread = JNIEnv2w_thread(env);
  w_clazz type;
  w_string typestring;

  if (!Type) {
    throwException(thread, clazzNullPointerException, NULL);
    return;
  }

  type = Class2clazz(Type);

  typestring = clazz2desc(type);

  if(!typestring){
    return;
  }

  setReferenceField(this, newStringInstance(typestring), F_ObjectStreamField_typeString);
  deregisterString(typestring);


  setBooleanField(this, F_ObjectStreamField_isPrimitive, WONKA_TRUE);

  if (type->dims) {
    setBooleanField(this, F_ObjectStreamField_isPrimitive, WONKA_FALSE);
    setCharacterField(this, F_ObjectStreamField_code, '[');
  }
  else if (type == clazz_byte) {
    setCharacterField(this, F_ObjectStreamField_code, 'B');
  }
  else if (type == clazz_char) {
    setCharacterField(this, F_ObjectStreamField_code, 'C');
  }
  else if (type == clazz_double) {
    setCharacterField(this, F_ObjectStreamField_code, 'D');
  }
  else if (type == clazz_float) {
    setCharacterField(this, F_ObjectStreamField_code, 'F');
  }
  else if (type == clazz_int) {
    setCharacterField(this, F_ObjectStreamField_code, 'I');
  }
  else if (type == clazz_long) {
    setCharacterField(this, F_ObjectStreamField_code, 'J');
  }
  else if (type == clazz_short) {
    setCharacterField(this, F_ObjectStreamField_code, 'S');
  }
  else if (type == clazz_boolean) {
    setCharacterField(this, F_ObjectStreamField_code, 'Z');
  }
  else {
    /*
    ** Must be another class or Interface type.
    */
    setBooleanField(this, F_ObjectStreamField_isPrimitive, WONKA_FALSE);
    setCharacterField(this, F_ObjectStreamField_code, 'L');
  }

  woempa(6, "fieldname is '%w', type is Class '%k', type code set at '%c'.\n", String2string(String), type, getCharacterField(this, F_ObjectStreamField_code));

  if (mustBeInitialized(clazzField) == CLASS_LOADING_FAILED) {
    return;
  }

  setReferenceField(this, Type, F_ObjectStreamField_type);
  setReferenceField(this, String, F_ObjectStreamField_name);
}


